### DISCUSSION TITLE
Self signed certificate when using non default workdir

### DISCUSSION THREAD

**Discussion by gaeljw** [renovatebot/renovate#23859] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.27.2

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Hello folks,

I've been using Renovate self-hosted (running through a container) for quite some time successfully, including with the configuration to our private registry using a self-signed certificate.

The command I was using so far is the following:
```bash
podman run --rm \
  -u "root:root" \
  -e "LOG_LEVEL=debug" \
  -e "RENOVATE_BASE_DIR=./renovate" \
  -e "NODE_EXTRA_CA_CERTS=/usr/src/app/certs.pem" \
  -e "RENOVATE_PLATFORM=gitlab" \
  -e "RENOVATE_TOKEN=..." \
  -e "RENOVATE_ENDPOINT=http://gitlab..../api/v4" \
  -v "config.js:/usr/src/app/config.js" \
  -v "/somepath/certs.pem:/usr/src/app/certs.pem" \
  renovate/renovate:36.27.2
```

The `config.js` file contains also a `npmrc` attribute with a reference to the `cafile`.

Especially, the certs to be used were referenced in `NODE_EXTRA_CA_CERTS` and mounted in `/usr/src/app/certs.pem`.

-----

**Now, I wanted to move this to Gitlab CI and there are 2 constraints that prevent the SSL config to be applied properly:**
- the workdir is forced to something else than `/usr/src/app`: it's something like `/builds/myproject` 
- the container is not run as root anymore but as the default user and thus I cannot mount files in `/usr/src/app`

This is equivalent as running the following:
```bash
podman run --rm \
  -e "LOG_LEVEL=debug" \
  -e "RENOVATE_BASE_DIR=./renovate" \
  -e "NODE_EXTRA_CA_CERTS=/build/my-project/certs.pem" \
  -e "RENOVATE_PLATFORM=gitlab" \
  -e "RENOVATE_TOKEN=..." \
  -e "RENOVATE_ENDPOINT=http://gitlab..../api/v4" \
  -v "config.js:/build/my-project/config.js" \
  -v "/somepath/certs.pem:/build/my-project/certs.pem" \
  -w "/build/my-project"
  renovate/renovate:36.27.2
```

When running this way, NPM seems to not completely work fine. What I observe is that Merge Requests are created for the `package.json` file but the `package-lock.json` file is not updated and an error is added in comments of the Merge Request:
```
npm ERR! code UNABLE_TO_VERIFY_LEAF_SIGNATURE
npm ERR! errno UNABLE_TO_VERIFY_LEAF_SIGNATURE
npm ERR! request to https://**redacted**@babel%2fcli failed, reason: unable to verify the first certificate

npm ERR! A complete log of this run can be found in: /builds/my-project/renovate/cache/others/npm/_logs/2023-08-14T15_10_22_920Z-debug-0.log
```

-----

**Said differently: the Renovate image seems to not work properly for NPM when running in a workdir which is not `/usr/src/app`.
If, in order to troubleshoot, I run the image with root user and copy the certs in `/usr/src/app/certs.pem`, everything works back fine.**

Am I missing something? Any help appreciated.

I guess if the `/usr/src/app` was not owned by root but by the default user, this would be enough to make it work as I could still mount the file in the directory expected.

----

Additional note: the issue is when Renovate runs `npm install --package-lock-only --no-audit --ignore-scripts` command.

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by viceice** [comment]
you currently need to configure `NODE_EXTRA_CA_CERTS` ad custom ENV var on renovate, renovate will pass it to child process.

**Reply by gaeljw** [reply]
This is what I'm already doing but it seems to not be enough when the referenced file is not in `/usr/src/app`.

Now that I think about it again, could it be that the file needs extra permission? I'll try by `chmod a+rwx` the file.

**Reply by viceice** [reply]
you probably need to remove the npmrc ca config .

have you added the ENV var here? otherwise it's not passed to child process.
https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Reply by gaeljw** [reply]
Ah, indeed I think I might need to drop the `npmrc` as it's being inherited from a preset with a wrong value. I will check that tomorrow.

Didn't use the `customEnvVariables` yet, will give it a try as well.

Thanks for your time, I'll keep you posted

**Reply by gaeljw** [reply]
Ok this was it: the `npmrc` attribute defined in a preset I used was used instead of the one I defined at a lower level.

Actually my setup is the following:
- project level: `renovate.json` with `"extends": ["local>myteam/renovate-config"]`
- team level (repo `myteam/renovate-config`): `default.json` with :
  - `"extends": [ "local>common/company-default-renovate-config"]`
  - a `"npmrc"` value (with correct path to `cafile`)
 - company level (repo `common/company-default-renovate-config`): `default.json` with a `"npmrc"` but wrong value for `cafile`

Removing the `"npmrc"` at the company level did the trick.

That sounds somehow like a bug to me though: I was expecting the team level config to take precedence.

Any opinion on this @viceice ?

(I didn't need the `customEnvVariables`)

**Reply by gaeljw** [reply]
Well, actually, entirely removing the `cafile` in `npmrc` works as well. Even easier. Thanks @viceice :)

**Reply by viceice** [reply]
I'll do a renovate update soon, so you don't need the custom environment variable inside renovate config at all.

**Reply by viceice** [reply, marked as answer]
instead of `NODE_EXTRA_CA_CERTS` you should use `SSL_CERT_FILE` as we set `NODE_OPTIONS=--use-openssl-ca` internally on renovate images. This will also allow other tools to use your custom CA.

https://github.com/containerbase/base/blob/main/docs/custom-root-ca.md

**Reply by viceice** [reply]
- https://github.com/renovatebot/renovate/pull/23887

**Reply by gaeljw** [reply]
I tested out with `SSL_CERT_FILE` in latest Renovate version instead of `NODE_EXTRA_CA_CERTS` and seems to work fine. Thanks again.

(For anyone reading this thread, in the end I had to use the `customEnvVariables` when using `NODE_EXTRA_CA_CERTS`  previously, but no more necessary now)
